File: tests/test_database.py
"""Tests for src/database.py"""
import json
import sqlite3
import tempfile
from pathlib import Path

import pytest

from src.database import FPLDatabase
from src.models import (
    Fixture,
    Gameweek,
    LiveGameweekStats,
    Player,
    PlayerHistory,
    PlayerHistoryPast,
    Team,
)
from src.transform import (
    transform_bootstrap,
    transform_element_summary,
    transform_event_live,
    transform_fixtures,
)

FIXTURES = Path(__file__).parent / "fixtures"


def load(filename: str) -> dict | list:
    return json.loads((FIXTURES / filename).read_text())


@pytest.fixture
def db(tmp_path):
    """Provide a fresh in-memory-style database for each test."""
    db_file = str(tmp_path / "test.db")
    database = FPLDatabase(db_file)
    database.initialize_schema()
    yield database
    database.close()


class TestSchemaInit:
    def test_tables_created(self, db):
        cur = db._conn.execute(
            "SELECT name FROM sqlite_master WHERE type='table' ORDER BY name"
        )
        tables = {row[0] for row in cur.fetchall()}
        expected = {
            "teams", "gameweeks", "players", "player_history",
            "player_history_past", "fixtures", "live_gameweek_stats", "scrape_log",
        }
        assert expected.issubset(tables)

    def test_idempotent(self, db):
        """Calling initialize_schema twice should not raise."""
        db.initialize_schema()


class TestUpsertTeams:
    def test_insert(self, db):
        data = load("bootstrap_static.json")
        teams, _, _ = transform_bootstrap(data)
        count = db.upsert_teams(teams)
        assert count == 1
        cur = db._conn.execute("SELECT fpl_id, name FROM teams WHERE fpl_id = 1")
        row = cur.fetchone()
        assert row["name"] == "Arsenal"

    def test_upsert_updates_existing(self, db):
        data = load("bootstrap_static.json")
        teams, _, _ = transform_bootstrap(data)
        db.upsert_teams(teams)
        # Modify and upsert again
        teams[0] = Team(
            fpl_id=1, name="Arsenal FC", short_name="ARS",
            strength=5, strength_overall_home=None, strength_overall_away=None,
            strength_attack_home=None, strength_attack_away=None,
            strength_defence_home=None, strength_defence_away=None, pulse_id=1,
        )
        db.upsert_teams(teams)
        cur = db._conn.execute("SELECT name FROM teams WHERE fpl_id = 1")
        assert cur.fetchone()["name"] == "Arsenal FC"


class TestUpsertPlayers:
    def test_insert_and_query(self, db):
        data = load("bootstrap_static.json")
        _, _, players = transform_bootstrap(data)
        db.upsert_players(players)
        ids = db.get_all_player_fpl_ids()
        assert 318 in ids

    def test_upsert_updates_cost(self, db):
        data = load("bootstrap_static.json")
        _, _, players = transform_bootstrap(data)
        db.upsert_players(players)
        # Simulate a price change
        p = players[0]
        updated = Player(
            fpl_id=p.fpl_id, first_name=p.first_name, second_name=p.second_name,
            web_name=p.web_name, team_fpl_id=p.team_fpl_id,
            element_type=p.element_type, status=p.status, code=p.code,
            now_cost=135,  # price rise
            cost_change_start=p.cost_change_start, cost_change_event=5,
            chance_of_playing_this_round=p.chance_of_playing_this_round,
            chance_of_playing_next_round=p.chance_of_playing_next_round,
            total_points=p.total_points, event_points=p.event_points,
            points_per_game=p.points_per_game, form=p.form,
            selected_by_percent=p.selected_by_percent,
            transfers_in=p.transfers_in, transfers_out=p.transfers_out,
            transfers_in_event=p.transfers_in_event,
            transfers_out_event=p.transfers_out_event,
            minutes=p.minutes, goals_scored=p.goals_scored, assists=p.assists,
            clean_sheets=p.clean_sheets, goals_conceded=p.goals_conceded,
            own_goals=p.own_goals, penalties_saved=p.penalties_saved,
            penalties_missed=p.penalties_missed, yellow_cards=p.yellow_cards,
            red_cards=p.red_cards, saves=p.saves, bonus=p.bonus, bps=p.bps,
            influence=p.influence, creativity=p.creativity,
            threat=p.threat, ict_index=p.ict_index, starts=p.starts,
            expected_goals=p.expected_goals, expected_assists=p.expected_assists,
            expected_goal_involvements=p.expected_goal_involvements,
            expected_goals_conceded=p.expected_goals_conceded,
            xgp=p.xgp, xap=p.xap, xgip=p.xgip,
            news=p.news, news_added=p.news_added, squad_number=p.squad_number,
            photo=p.photo,
        )
        db.upsert_players([updated])
        cur = db._conn.execute("SELECT now_cost FROM players WHERE fpl_id = 318")
        assert cur.fetchone()["now_cost"] == 135


class TestUpsertPlayerHistory:
    def test_insert(self, db):
        data = load("bootstrap_static.json")
        _, _, players = transform_bootstrap(data)
        db.upsert_players(players)

        summary_data = load("element_summary_318.json")
        history, history_past = transform_element_summary(318, summary_data)
        db.upsert_player_history(history)
        db.upsert_player_history_past(history_past)

        cur = db._conn.execute(
            "SELECT total_points FROM player_history WHERE player_fpl_id = 318 AND gameweek_fpl_id = 24"
        )
        assert cur.fetchone()["total_points"] == 14

        cur = db._conn.execute(
            "SELECT season_name FROM player_history_past WHERE player_fpl_id = 318"
        )
        assert cur.fetchone()["season_name"] == "2023/24"

    def test_upsert_is_idempotent(self, db):
        summary_data = load("element_summary_318.json")
        history, _ = transform_element_summary(318, summary_data)
        db.upsert_player_history(history)
        db.upsert_player_history(history)  # second time should not raise
        cur = db._conn.execute("SELECT count(*) as n FROM player_history WHERE player_fpl_id = 318")
        assert cur.fetchone()["n"] == 1


class TestUpsertFixtures:
    def test_insert(self, db):
        fixtures_data = load("fixtures.json")
        fixtures = transform_fixtures(fixtures_data)
        db.upsert_fixtures(fixtures)
        cur = db._conn.execute("SELECT fpl_id FROM fixtures WHERE fpl_id = 1")
        assert cur.fetchone() is not None


class TestUpsertLiveStats:
    def test_insert(self, db):
        live_data = load("event_25_live.json")
        rows = transform_event_live(25, live_data)
        db.upsert_live_stats(rows)
        cur = db._conn.execute(
            "SELECT total_points FROM live_gameweek_stats "
            "WHERE player_fpl_id = 318 AND gameweek_fpl_id = 25"
        )
        assert cur.fetchone()["total_points"] == 12


class TestGameweekQueries:
    def test_get_current_gameweek(self, db):
        data = load("bootstrap_static.json")
        _, gameweeks, _ = transform_bootstrap(data)
        db.upsert_gameweeks(gameweeks)
        gw = db.get_current_gameweek()
        assert gw is not None
        assert gw["fpl_id"] == 25

    def test_get_current_gameweek_returns_none_when_empty(self, db):
        assert db.get_current_gameweek() is None


class TestScrapeLog:
    def test_start_and_finish(self, db):
        db.start_scrape_log("run-abc", "full_sync", None, "2025-01-01T00:00:00Z")
        db.finish_scrape_log(
            "run-abc", "success", 700, 750, 0, "2025-01-01T01:30:00Z"
        )
        cur = db._conn.execute(
            "SELECT status, players_scraped FROM scrape_log WHERE run_id = 'run-abc'"
        )
        row = cur.fetchone()
        assert row["status"] == "success"
        assert row["players_scraped"] == 700

File: tests/test_database.ts
import { mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { afterEach, describe, expect, it } from "vitest";

import { FPLDatabase } from "../src/database.ts";
import { Fixture, Gameweek, Team } from "../src/models.ts";
import {
  transform_bootstrap,
  transform_element_summary,
  transform_event_live,
  transform_fixtures,
} from "../src/transform.ts";

const FIXTURES_DIR = join(dirname(fileURLToPath(import.meta.url)), "fixtures");
const TEMP_DIRECTORIES: string[] = [];
const OPEN_DATABASES: FPLDatabase[] = [];

function loadFixture<T>(filename: string): T {
  return JSON.parse(readFileSync(join(FIXTURES_DIR, filename), "utf8")) as T;
}

function createDatabase(): FPLDatabase {
  const tempDirectory = mkdtempSync(join(tmpdir(), "fpl-database-"));
  TEMP_DIRECTORIES.push(tempDirectory);

  const database = new FPLDatabase(join(tempDirectory, "test.db"));
  OPEN_DATABASES.push(database);
  database.initializeSchema();
  return database;
}

afterEach(() => {
  for (const database of OPEN_DATABASES.splice(0)) {
    try {
      database.close();
    } catch {
      // Tests may already have closed the connection explicitly.
    }
  }

  for (const tempDirectory of TEMP_DIRECTORIES.splice(0)) {
    rmSync(tempDirectory, { force: true, recursive: true });
  }
});

describe("src/database.ts", () => {
  it("creates the full schema, keeps initialization idempotent, and preserves SQLite pragmas", () => {
    const db = createDatabase();

    db.initializeSchema();

    const tables = db._conn
      .prepare(
        "SELECT name FROM sqlite_master WHERE type = 'table' ORDER BY name",
      )
      .all() as Array<{ name: string }>;
    const tableNames = tables.map((row) => row.name);

    expect(tableNames).toEqual(
      expect.arrayContaining([
        "fixtures",
        "gameweeks",
        "live_gameweek_stats",
        "player_history",
        "player_history_past",
        "players",
        "scrape_log",
        "teams",
      ]),
    );
    expect(db._conn.pragma("journal_mode", { simple: true })).toBe("wal");
    expect(db._conn.pragma("foreign_keys", { simple: true })).toBe(1);
  });

  it("upserts bootstrap rows and answers the gameweek/player helpers", () => {
    const db = createDatabase();
    const bootstrap = loadFixture<Record<string, unknown>>(
      "bootstrap_static.json",
    );
    const [teams, gameweeks, players] = transform_bootstrap(bootstrap);
    const baseCurrentGameweek =
      gameweeks.find((gameweek) => gameweek.fpl_id === 25) ?? gameweeks[0]!;
    const baseNextGameweek =
      gameweeks.find(
        (gameweek) => gameweek.fpl_id !== baseCurrentGameweek.fpl_id,
      ) ?? gameweeks[0]!;

    expect(db.upsertTeams(teams)).toBe(1);
    expect(db.upsertGameweeks(gameweeks)).toBe(2);
    expect(db.upsertPlayers(players)).toBe(1);

    const seededCurrentGameweek = new Gameweek(
      baseCurrentGameweek.fpl_id,
      baseCurrentGameweek.name,
      baseCurrentGameweek.deadline_time,
      baseCurrentGameweek.average_entry_score,
      baseCurrentGameweek.highest_score,
      baseCurrentGameweek.highest_scoring_entry,
      1,
      0,
      baseCurrentGameweek.is_finished,
      baseCurrentGameweek.chip_plays,
      baseCurrentGameweek.most_selected,
      baseCurrentGameweek.most_transferred_in,
      baseCurrentGameweek.most_captained,
      baseCurrentGameweek.most_vice_captained,
      baseCurrentGameweek.transfers_made,
      baseCurrentGameweek.scraped_at,
    );
    db.upsertGameweeks([seededCurrentGameweek]);

    const seededNextGameweek = new Gameweek(
      baseNextGameweek.fpl_id,
      baseNextGameweek.name,
      baseNextGameweek.deadline_time,
      baseNextGameweek.average_entry_score,
      baseNextGameweek.highest_score,
      baseNextGameweek.highest_scoring_entry,
      0,
      1,
      baseNextGameweek.is_finished,
      baseNextGameweek.chip_plays,
      baseNextGameweek.most_selected,
      baseNextGameweek.most_transferred_in,
      baseNextGameweek.most_captained,
      baseNextGameweek.most_vice_captained,
      baseNextGameweek.transfers_made,
      baseNextGameweek.scraped_at,
    );
    db.upsertGameweeks([seededNextGameweek]);

    const updatedTeam = new Team(
      1,
      "Arsenal FC",
      "ARS",
      3,
      5,
      1200,
      1190,
      1180,
      1170,
      1160,
      1150,
      1,
      teams[0]?.scraped_at,
    );
    db.upsertTeams([updatedTeam]);

    const currentGameweek = db.getCurrentGameweek();
    const nextGameweek = db.getNextGameweek();
    const gameweekById = db.getGameweekById(25);

    expect(currentGameweek).not.toBeNull();
    expect(currentGameweek?.fpl_id).toBe(baseCurrentGameweek.fpl_id);
    expect(nextGameweek?.fpl_id).toBe(baseNextGameweek.fpl_id);
    expect(gameweekById?.name).toBe(baseCurrentGameweek.name);
    expect(db.getAllPlayerFplIds()).toEqual([318]);
    const teamRow = db._conn
      .prepare("SELECT name FROM teams WHERE fpl_id = 1")
      .get() as { name: string } | undefined;
    expect(teamRow?.name).toBe("Arsenal FC");
    expect(
      db._conn
        .prepare(
          "SELECT now_cost, xgp, tackles FROM players WHERE fpl_id = 318",
        )
        .get(),
    ).toMatchObject({
      now_cost: players[0]!.now_cost,
      xgp: players[0]!.xgp,
      tackles: players[0]!.tackles,
    });
  });

  it("upserts player history, past history, fixtures, and live stats idempotently from real payloads", () => {
    const db = createDatabase();
    const [history, historyPast] = transform_element_summary(
      318,
      loadFixture<Record<string, unknown>>("element_summary_318.json"),
    );
    const fixtures = transform_fixtures(loadFixture<unknown>("fixtures.json"));
    const liveStats = transform_event_live(
      25,
      loadFixture<Record<string, unknown>>("event_25_live.json"),
    );

    expect(db.upsertPlayerHistory(history)).toBe(1);
    expect(db.upsertPlayerHistory(history)).toBe(1);
    expect(db.upsertPlayerHistoryPast(historyPast)).toBe(1);
    expect(db.upsertFixtures(fixtures)).toBe(1);
    expect(db.upsertLiveGameweekStats(liveStats)).toBe(1);

    expect(
      db._conn
        .prepare(
          "SELECT COUNT(*) AS count FROM player_history WHERE player_fpl_id = 318",
        )
        .get(),
    ).toMatchObject({ count: 1 });
    expect(
      db._conn
        .prepare(
          "SELECT season_name FROM player_history_past WHERE player_fpl_id = 318",
        )
        .get(),
    ).toMatchObject({ season_name: "2023/24" });
    expect(
      db._conn.prepare("SELECT stats FROM fixtures WHERE fpl_id = 1").get(),
    ).toMatchObject({
      stats: Fixture.fromDict(
        loadFixture<Record<string, unknown>[]>("fixtures.json")[0] as Record<
          string,
          unknown
        >,
      ).stats,
    });
    expect(
      db._conn
        .prepare(
          "SELECT total_points, explain FROM live_gameweek_stats WHERE player_fpl_id = 318 AND gameweek_fpl_id = 25",
        )
        .get(),
    ).toMatchObject({
      total_points: 12,
      explain: liveStats[0]?.explain,
    });
    expect(db.getActivePlayerIdsInGw(25)).toEqual([318]);
  });

  it("tracks scrape logs and returns the most recent successful run for a mode", () => {
    const db = createDatabase();

    db.startScrapeLog("run-error", "full_sync", null, "2025-01-01T00:00:00Z");
    db.finishScrapeLog(
      "run-error",
      "error",
      10,
      20,
      1,
      "2025-01-01T00:15:00Z",
      "boom",
    );
    db.startScrapeLog("run-success", "full_sync", 25, "2025-01-01T01:00:00Z");
    db.finishScrapeLog(
      "run-success",
      "success",
      700,
      750,
      0,
      "2025-01-01T01:30:00Z",
    );

    const latestSuccess = db.getLastSuccessfulScrape("full_sync");

    expect(latestSuccess).not.toBeNull();
    expect(latestSuccess).toMatchObject({
      run_id: "run-success",
      gameweek_fpl_id: 25,
      status: "success",
      players_scraped: 700,
      requests_made: 750,
      errors_encountered: 0,
      error_detail: null,
    });
    expect(
      db._conn
        .prepare("SELECT status FROM scrape_log WHERE run_id = ?")
        .get("run-error"),
    ).toMatchObject({ status: "error" });
  });

  it("closes the underlying SQLite connection", () => {
    const db = createDatabase();

    db.close();

    expect(() => db._conn.prepare("SELECT 1").get()).toThrow(/not open/i);
  });
});
